File: services/tms/internal/api/handlers/googlemaps.go
package handlers

import (
	"errors"
	"net/http"

	"github.com/emoss08/trenova/internal/api/helpers"
	"github.com/emoss08/trenova/internal/infrastructure/config"
	"github.com/emoss08/trenova/pkg/googlemaps"
	"github.com/gin-gonic/gin"
	"go.uber.org/fx"
)

type GoogleMapsHandlerParams struct {
	fx.In

	AutoCompleteService *googlemaps.AutoCompleteService
	Config              *config.Config
	ErrorHandler        *helpers.ErrorHandler
}

type GoogleMapsHandler struct {
	autoCompleteService *googlemaps.AutoCompleteService
	config              *config.Config
	eh                  *helpers.ErrorHandler
}

func NewGoogleMapsHandler(p GoogleMapsHandlerParams) *GoogleMapsHandler {
	return &GoogleMapsHandler{
		autoCompleteService: p.AutoCompleteService,
		config:              p.Config,
		eh:                  p.ErrorHandler,
	}
}

func (h *GoogleMapsHandler) RegisterRoutes(rg *gin.RouterGroup) {
	api := rg.Group("/google-maps/")
	api.POST("/autocomplete/", h.autocomplete)
	api.GET("/api-key/", h.getAPIKey)
}

func (h *GoogleMapsHandler) autocomplete(c *gin.Context) {
	var req googlemaps.AutoCompleteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		h.eh.HandleError(c, err)
		return
	}

	response, err := h.autoCompleteService.GetPlaceDetails(c.Request.Context(), &req)
	if err != nil {
		h.eh.HandleError(c, err)
		return
	}

	c.JSON(http.StatusOK, response)
}

func (h *GoogleMapsHandler) getAPIKey(c *gin.Context) {
	apiKey := h.config.Google.APIKey
	if apiKey == "" {
		h.eh.HandleError(c, errors.New("API key is not set"))
		return
	}

	c.JSON(http.StatusOK, gin.H{"apiKey": apiKey})
}

File: client/apps/web/src/services/google-maps.ts
import { api } from "@trenova/shared/lib/api";
import { safeParse } from "@trenova/shared/lib/parse";
import {
  integrationRuntimeConfigResponseSchema,
  type IntegrationRuntimeConfigResponse,
} from "@/types/integration";
import {
  type AutocompleteLocationRequest,
  type AutocompleteLocationResult,
  autocompleteLocationResultSchema,
} from "@/types/google-maps";

export class GoogleMapsService {
  public async getAPIKey(): Promise<IntegrationRuntimeConfigResponse> {
    const response = await api.get("/integrations/GoogleMaps/runtime-config/");
    return safeParse(
      integrationRuntimeConfigResponseSchema,
      response,
      "Google Maps Runtime Config",
    );
  }

  public async autocomplete(req: AutocompleteLocationRequest) {
    const response = await api.post<AutocompleteLocationResult>(`/google-maps/autocomplete/`, req);

    return safeParse(autocompleteLocationResultSchema, response, "AutocompleteLocationResult");
  }
}
